exchanges/kucoin/kucoin_futures.go

```diff
@@ -0,0 +1,858 @@
+		HasMore bool                  `json:"hasMore"`
+	}{}
+	return resp.List, ku.SendHTTPRequest(ctx, exchange.RestFutures, defaultFuturesEPL, common.EncodeURLValues(kucoinFuturesPremiumIndex, params), &resp)
+}
+
+// GetFuturesCurrentFundingRate get current funding rate
+func (ku *Kucoin) GetFuturesCurrentFundingRate(ctx context.Context, symbol string) (*FuturesFundingRate, error) {
+	if symbol == "" {
+		return nil, errors.New("symbol can't be empty")
+	}
+	var resp *FuturesFundingRate
+	return resp, ku.SendHTTPRequest(ctx, exchange.RestFutures, defaultFuturesEPL, fmt.Sprintf(kucoinFuturesFundingRate, symbol), &resp)
+}
+
+// GetFuturesServerTime get server time
+func (ku *Kucoin) GetFuturesServerTime(ctx context.Context, symbol string) (time.Time, error) {
+	resp := struct {
+		Data kucoinTimeMilliSec `json:"data"`
+		Error
+	}{}
+	return resp.Data.Time(), ku.SendHTTPRequest(ctx, exchange.RestFutures, defaultFuturesEPL, kucoinFuturesServerTime, &resp)
+}
+
+// GetFuturesServiceStatus get service status
+func (ku *Kucoin) GetFuturesServiceStatus(ctx context.Context, symbol string) (*FuturesServiceStatus, error) {
+	var resp *FuturesServiceStatus
+	return resp, ku.SendHTTPRequest(ctx, exchange.RestFutures, defaultFuturesEPL, kucoinFuturesServiceStatus, &resp)
+}
+
+// GetFuturesKline get contract's kline data
+func (ku *Kucoin) GetFuturesKline(ctx context.Context, granularity, symbol string, from, to time.Time) ([]FuturesKline, error) {
+	if granularity == "" {
+		return nil, errors.New("granularity can not be empty")
+	}
+	if !common.StringDataContains(validGranularity, granularity) {
+		return nil, errors.New("invalid granularity")
+	}
+	params := url.Values{}
+	// The granularity (granularity parameter of K-line) represents the number of minutes, the available granularity scope is: 1,5,15,30,60,120,240,480,720,1440,10080. Requests beyond the above range will be rejected.
+	params.Set("granularity", granularity)
+	if symbol == "" {
+		return nil, errors.New("symbol can't be empty")
+	}
+	params.Set("symbol", symbol)
+	if !from.IsZero() {
+		params.Set("from", strconv.FormatInt(from.UnixMilli(), 10))
+	}
+	if !to.IsZero() {
+		params.Set("to", strconv.FormatInt(to.UnixMilli(), 10))
+	}
+	var resp [][]float64
+
+	err := ku.SendHTTPRequest(ctx, exchange.RestFutures, defaultFuturesEPL, common.EncodeURLValues(kucoinFuturesKline, params), &resp)
+	if err != nil {
+		return nil, err
+	}
+	kline := make([]FuturesKline, len(resp))
+	for i := range resp {
+		kline[i] = FuturesKline{
+			StartTime: time.UnixMilli(int64(resp[i][0])),
+			Open:      resp[i][1],
+			High:      resp[i][2],
+			Low:       resp[i][3],
+			Close:     resp[i][4],
+			Volume:    resp[i][5],
+		}
+	}
+	return kline, nil
+}
+
+// PostFuturesOrder used to place two types of futures orders: limit and market
+func (ku *Kucoin) PostFuturesOrder(ctx context.Context, clientOID, side, symbol,
+	orderType, remark,
+	stop, stopPriceType, stopPrice, timeInForce string, size, price,
```

Comment:
You have some inconsistencies when it comes to prices and strings. For futures you require strings for prices but in SPOT you ask for float64 and convert to string